### ISSUE TITLE
Quality control issues with large datasets

### ISSUE THREAD

**Issue by janbrogger** [bwrc/ctap#9]
We're planning on using CTAP for around 10,000 clinical EEGs. Quality control becomes a big issue then. Are you planning any features that can help with this? Some of the features I just submitted in a pull request are intended to help with this.

Matlab isn't a good tool for database-like functionality. We need to simultaneously have a birds-eye view but also the ability to look at a single EEG in detail if needed. A birds eye view is necessary so you can know at a glance the result of your whole set things such as:
- Population step 0: % of EEGs that failed their CTAP run
- Population step 1: histograms and other descriptives of EEG length, number of channels, sampling rate so you can set thresholds to discard obviously "garbage" EEGs (e.g. 5-second clinical EEGs)
- Population step 2: badness descriptives, number of channels rejected etc, % rejected etc. This is to capture any obvious outliers that shouldn't be there at this point in the pipeline.
- Population step 3: distribution of various QEEG measures. This is to capture any obvious outliers that shouldn't be there at this point in the pipeline.

On a single-EEG level, it is currently a little difficult to track the processing steps. Our goal for this particular project is to verify that obvious artefacts and drowsy/sleepy EEG hasn't been included in subsequent steps in the pipeline. The ctap_results\ subdirectories are now ordered by step name so we have:

1. CTAP_blink2event
2. CTAP_detect_bad_channels
3. CTAP_detect_bad_comps
4. CTAP_fir_filter
5. CTAP_fir_notchfilter
6. CTAP_peek_data
7. CTAP_reject_data
8. CTAP_run_ica

How can we structure our CTAP output so you can follow one EEG through the steps? Any hints - can we number the output steps in any way? 

Ideally, you should be able to document the whole thing as PDF-file (one PDF for the population level, one PDF per single EEG). The ideal is something like Stata 15's "dyndoc" which basically is a text report interspersed with code. See here: https://www.stata.com/new-in-stata/markdown/  Would you be unhappy if we wrote code for Stata that took CTAP code/features to generate a report? E.g. ~\CTAP\ctap\stata-src\generate-report.do ?

There's a limit to how far you can push Matlab to do some of these things. Would you be happy with us adding database-like features to CTAP, e.g. using far more SqLite?

**Comment by zenBen** [comment]
Hei Jan

only a quick comment possible right now (it has been a busy week).
I like the direction of your thinking, and agree that meta-data has been neglected in CTAP development so far, thus creating a need for these features. We will never be unhappy if people are contributing good ideas, even better good code. 

Regarding our options to respond to your questions with action, I'll discuss with my co-author Jussi about how to address your issues, asap. I will try to give such answers as I can before the end of next week.

**Comment by jutako** [comment]
Regarding tracking what has been done in which order you have two options:
1. add a rolling index to you stepset names (stepSet(i).id) as has been done in runctap_minimal.m
2. use the branched pipe example (runctap_manu_branch.m) to organize stepsets in larger chunks such as "data loading", "pre processing", "feature extraction".

Approached 1. and 2. can also be combined. Without explicit hacks source data for a given stepset is the output of previous stepset. Order is indexed by i in stepSet(i).

I agree with Ben that any contribution is welcome. If I was to decide, I'd use R Markdown or Python Jupyter notebooks/similar to generate automatic reports. That said, commercial software is fine as well. After all it is the usefulness of the reports that matters.

**Comment by jutako** [comment]
I and Ben had a discussion about this and we came up with the following scheme of collecting "operational feature" data, i.e., data that describes what happened as the pipe was run.

## Design principles for operational feature export

- CTAP_*() function should only define what features to store and provide these in a simple format.
- CTAP_*() could report the features as Matlab struct with alphanumeric fieldnames and scalar or string data
- users could add features or make their own CTAP_*() that exports what they want
- some myReport() -style function (different name!) should take this info, add pipeline metadata and store
- pipeline metadata include at least::

1. casename
2. step set savepath (position in a branched pipe)
3. step set
4. step fun number
5. CTAP_*() function name, possibly operation mode padded if needed
6. timestamp (for debugging and conflict resolution purposes)

- operational features will be stored as:

1. feature name
2. feature value

- in other words: a long format dataset
- first version csv, possibly SQL database later. One file per call to pipeline looper, no attempts to update data at this point. Like log file but for the machine.
- Summary reports could be based on data from this file and implemented using any suitable software.

Do you think something along these lines could be helpful? We are interested in implementing the basic functionality, but providing all possible features for all possible CTAP_*() functions remains a longer term effort. Maybe more features should be added as the need arises.

**Comment by zenBen** [comment]
@janbrogger Hei Jan, have you any further comment on this matter? We have worked to update the public CTAP repo with all upgrades made since the paper was published, which should provide a few more features and bug fixes, as well as pulling in your contribution. As mentioned in the [latest issue](https://github.com/bwrc/ctap/issues/10), we advise to test your code against the dev branch of this repo, so that we can update the master branch without breaking your analyses.

I would also be interested to hear more about how your analysis is going, and what you think about our proposals above: this is by far the largest work CTAP has been used for, so it's valuable for us to know how it's going.
